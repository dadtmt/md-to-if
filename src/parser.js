import SimpleMarkdown from 'simple-markdown'

// String -> [String]
export const matchBraces = source => /^\{([\s\S]*?)\}/.exec(source)

// String -> [String]
export const matchBracketPipe = source => /^\[([\s\S]*?)\|/.exec(source)

// String -> [String]
export const matchPipeBracket = source => /^\|([\s\S]*?)\]/.exec(source)

const command = {
  order: 0,
  match: matchBraces,
  parse: (capture, parse, state) => ({ content: parse(capture[1], state) }),
}

const trueCaseContent = {
  order: 0,
  match: matchBracketPipe,
  parse: (capture, parse, state) => ({ content: parse(capture[1], state) }),
}

const falseCaseContent = {
  order: 0,
  match: matchPipeBracket,
  parse: (capture, parse, state) => ({ content: parse(capture[1], state) }),
}

// String -> [Content]
const parser = SimpleMarkdown.parserFor({
  ...SimpleMarkdown.defaultRules,
  command,
  trueCaseContent,
  falseCaseContent,
})

export default parser
