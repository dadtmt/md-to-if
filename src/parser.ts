import SimpleMarkdown, { SingleASTNode } from 'simple-markdown'

// String -> [String]
export const matchBraces = (source: string): RegExpExecArray | null =>
  /^\{([\s\S]*?)\}/.exec(source)

// String -> [String]
export const matchBracketPipe = (source: string): RegExpExecArray | null =>
  /^\[([\s\S]*?)\|/.exec(source)

// String -> [String]
export const matchPipeBracket = (source: string): RegExpExecArray | null =>
  /^\|([\s\S]*?)\]/.exec(source)

const parse: (capture: any, parse: any, state: any) => any = (
  capture,
  parse,
  state
) => ({ content: parse(capture[1], state) })

const command = {
  order: 0,
  match: matchBraces,
  parse
}

const trueCaseContent = {
  order: 0,
  match: matchBracketPipe,
  parse
}

const falseCaseContent = {
  order: 0,
  match: matchPipeBracket,
  parse
}

// String -> [Content]
const parser: (source: string) => SingleASTNode[] = SimpleMarkdown.parserFor({
  ...SimpleMarkdown.defaultRules,
  command,
  trueCaseContent,
  falseCaseContent
})

export default parser
