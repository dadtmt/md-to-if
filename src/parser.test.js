import parser, { matchBraces, matchSetter } from './parser'
import adventure from './adventure.md.js'

describe('matchBraces', () => {
  it('match braces', () => {
    const source = `{ show value } after
bla
bla
bla later some {show other} bla`
    expect(matchBraces(source)).toMatchSnapshot()
  })
})

describe('matchSetter', () => {
  it('match setter marks', () => {
    const source = `<! name value !> after
bla
bla
bla later some {show other} bla`
    expect(matchSetter(source)).toMatchSnapshot()
  })
})

describe('parser', () => {
  it('parse markdown', () => {
    expect(parser(adventure)).toMatchSnapshot()
  })
})
