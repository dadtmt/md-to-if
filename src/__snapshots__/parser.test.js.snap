// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matchBraces match braces 1`] = `
Array [
  "{ show value }",
  " show value ",
]
`;

exports[`parser parse markdown 1`] = `
Array [
  Object {
    "content": Array [
      Object {
        "content": "Choose your story",
        "type": "text",
      },
    ],
    "level": 1,
    "type": "heading",
  },
  Object {
    "content": Array [
      Object {
        "content": "This is a test adventure",
        "type": "text",
      },
      Object {
        "content": ", welcome",
        "type": "text",
      },
      Object {
        "content": ".",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "content": Array [
      Object {
        "content": "The space ship",
        "type": "text",
      },
    ],
    "level": 2,
    "type": "heading",
  },
  Object {
    "content": Array [
      Object {
        "content": "You are the brave pilot of this ship",
        "type": "text",
      },
      Object {
        "content": ".
The trip is long and boring",
        "type": "text",
      },
      Object {
        "content": ".",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "content": Array [
      Object {
        "content": "Are you going to ",
        "type": "text",
      },
      Object {
        "content": Array [
          Object {
            "content": "get food",
            "type": "text",
          },
        ],
        "target": "/cantina",
        "title": undefined,
        "type": "link",
      },
      Object {
        "content": " or to ",
        "type": "text",
      },
      Object {
        "content": Array [
          Object {
            "content": "get a rest",
            "type": "text",
          },
        ],
        "target": "/bedroom",
        "title": undefined,
        "type": "link",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "content": Array [
      Object {
        "content": "Cantina",
        "type": "text",
      },
    ],
    "level": 2,
    "type": "heading",
  },
  Object {
    "content": Array [
      Object {
        "content": " show playedCount ",
        "type": "text",
      },
    ],
    "type": "dynamic",
  },
  Object {
    "content": Array [
      Object {
        "content": " time you come here",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "content": Array [
      Object {
        "content": "The cantina is full with your crew having a good time",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "content": Array [
      Object {
        "content": "They are singing ",
        "type": "text",
      },
      Object {
        "content": "'oh my captain",
        "type": "text",
      },
      Object {
        "content": "' for the ",
        "type": "text",
      },
      Object {
        "content": Array [
          Object {
            "content": " show playedCount ",
            "type": "text",
          },
        ],
        "type": "dynamic",
      },
      Object {
        "content": " times",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "content": Array [
      Object {
        "content": "You pick up a sandwich an put it in your bag ",
        "type": "text",
      },
      Object {
        "content": Array [
          Object {
            "content": " set bag sandwich 1 ",
            "type": "text",
          },
        ],
        "type": "dynamic",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "content": Array [
      Object {
        "content": "You have ",
        "type": "text",
      },
      Object {
        "content": Array [
          Object {
            "content": " show bag sandwich ",
            "type": "text",
          },
        ],
        "type": "dynamic",
      },
      Object {
        "content": " with you",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
]
`;
